### PR TITLE
File upload optimization

### LINKED ISSUE #5829: Investigate UploadFile Optimization
- [ ] I have searched to see if a similar issue already exists.

**Is your feature request related to a problem? Please describe.**  
The UploadRoute uses `FileUpload` which under the hood uses a `SpooledTemporaryFile` [link](https://fastapi.tiangolo.com/tutorial/request-files/?h=upload#__tabbed_3_3). This will write the file to disk if its larger than a certain size. We then iterate over the file and write it to a separate location.

So we're writing the file twice. I wonder if we can just copy the file to a new location? Or iterate through the bytes of the file without having to write it to disk.

**Describe the solution you'd like**  

**Additional context**

### PR DESCRIPTION
**Pull request by freddyaboulton**
## Description

Closes #5829

The `/upload` route would read and write through each file twice. When you type a form as `UploadFile`, fastapi will parse all of the incoming bytes into an UploadFile (this is the first time we read/write the file). We then read through all of the file again (to compute the hash) and write it to a separate location. 

To only do it once, we can use our own multipart parser to parse the incoming request. This multipart parser is vendored from starlette but we made two modifications - 1) We use a NamedTemporaryFile instead of a SpooledTemporaryFile and 2) We compute the hash of the incoming bytes as we parse the incoming request.

Did a small test locally (to hold the upload speed constant) and the total time spent in the route after parsing the request is much smaller

### This branch
image

### V4
image

## 🎯 PRs Should Target Issues

Before your create a PR, please check to see if there is [an existing issue](https://github.com/gradio-app/gradio/issues) for this change. If not, please create an issue before you create this PR, unless the fix is very small. 

Not adhering to this guideline will result in the PR being closed. 

## Tests

1. PRs will only be merged if tests pass on CI. To run the tests locally, please set up [your Gradio environment locally](https://github.com/gradio-app/gradio/blob/main/CONTRIBUTING.md) and run the tests: `bash scripts/run_all_tests.sh`

2. You may need to run the linters: `bash scripts/format_backend.sh` and `bash scripts/format_frontend.sh`

### CODE REVIEW

File: gradio/route_utils.py [hunk 1]
```diff
@@ -264,3 +270,189 @@ def strip_url(orig_url: str) -> str:
...
+        *,
+        size: int | None = None,
+        filename: str | None = None,
+        headers: Headers | None = None,
+    ) -> None:
+        super().__init__(file, size=size, filename=filename, headers=headers)
+        self.sha = hashlib.sha1()
+
+
+class GradioMultiPartParser:
```
Comment (freddyaboulton) [inline gradio/route_utils.py:297]: This is vendored from starlette. It's also possible to do this via subclassing MultipartParser from starlette but that leaves us vulnerable to any changes starlette makes to that class.

File: test/test_gradio_component_cli.py [hunk 2]
```diff
@@ -60,7 +60,7 @@ def test_overwrite_deletes_previous_content(tmp_path):
     _create("MyGallery", tmp_path, template="Gallery", overwrite=True)
     _create("MySlider", tmp_path, template="Slider", overwrite=True)
     assert (tmp_path / "frontend" / "interactive" / "InteractiveSlider.svelte").exists()
-    assert not (tmp_path / "frontend" / "static" / "StaticGallery.svelte").exists()
+    assert not (tmp_path / "frontend" / "shared" / "Gallery.svelte").exists()
```
Comment (abidlabs) [inline test/test_gradio_component_cli.py:63]: Not related to this PR right?

**Comment by abidlabs** [comment]
Very cool PR @freddyaboulton! I tested a bunch of demos and they work great.

The one demo that isn't working (though I don't know the root cause) is `zip_files`. Seeing it error out both locally and on Spaces. 

```
Traceback (most recent call last):
  File "/home/user/.local/lib/python3.10/site-packages/gradio/queueing.py", line 407, in call_prediction
    output = await route_utils.call_process_api(
  File "/home/user/.local/lib/python3.10/site-packages/gradio/route_utils.py", line 232, in call_process_api
    output = await app.get_blocks().process_api(
  File "/home/user/.local/lib/python3.10/site-packages/gradio/blocks.py", line 1424, in process_api
    inputs = self.preprocess_data(fn_index, inputs, state)
  File "/home/user/.local/lib/python3.10/site-packages/gradio/blocks.py", line 1206, in preprocess_data
    processed_input.append(block.preprocess(inputs_cached))
  File "/home/user/.local/lib/python3.10/site-packages/gradio/components/file.py", line 197, in preprocess
    return [self._process_single_file(f, type=self.type, cache_dir=self.GRADIO_CACHE) for f in x]  # type: ignore
  File "/home/user/.local/lib/python3.10/site-packages/gradio/components/file.py", line 197, in <listcomp>
    return [self._process_single_file(f, type=self.type, cache_dir=self.GRADIO_CACHE) for f in x]  # type: ignore
  File "/home/user/.local/lib/python3.10/site-packages/gradio/components/file.py", line 152, in _process_single_file
    f["name"],
KeyError: 'name'


```

File: test/test_gradio_component_cli.py (hunk 2 above)
Comment (freddyaboulton) [reply test/test_gradio_component_cli.py:63]: I did this cause a test was failing after the front-end reorg.Not sure if it actually fixed the problem though.

**Comment by freddyaboulton** [comment]
Sweet - will look at that demo I thought it was working.

**Comment by freddyaboulton** [comment]
I think it's working now @abidlabs !

**Review by abidlabs — APPROVED** [review]
Everything works! LGTM & really nice perf boost 🔥

**Comment by freddyaboulton** [comment]
Thanks @abidlabs !

**Pull request merged**